app/src/main/java/net/squanchy/schedule/ScheduleService.kt

```diff
@@ -1,88 +1,100 @@
-        val uniqueTracks = scheduleView.map { it.events }
-            .flatten()
-            .mapNotNull { it.track }
-            .map { it.toTrack() }
-            .distinct()
-        filterScheduleRepository.allTracks = uniqueTracks
-    }
+    private fun Observable<List<FirestoreSchedulePage>>.filterByFavorites(onlyFavorites: Boolean) =
+        when {
+            onlyFavorites -> this.removeNonFavorites()
+            else -> this
+        }
 
-    private fun createSchedulePages(
-        schedulePages: List<FirestoreSchedulePage>,
-        timeZone: DateTimeZone,
-        onlyFavorites: Boolean,
-        selectedTracks: Set<Track>,
-        checksum: Checksum
-    ) = schedulePages.map { schedulePage ->
-        SchedulePage(
-            schedulePage.day.id,
-            LocalDate(schedulePage.day.date),
-            // TODO we don't have favourites yet (they're all "false")
-            schedulePage.events.map { it.toEvent(checksum, timeZone) }
-                .sortedBy { it.startTime }
-                .filterOnlyFavorites(onlyFavorites)
-                .filterByTracks(selectedTracks)
+    private fun Observable<List<FirestoreSchedulePage>>.removeNonFavorites(): Observable<List<FirestoreSchedulePage>> =
+        map { pages: List<FirestoreSchedulePage> -> pages.filterPagesEvents { it.isFavorite() } }
+
+    private fun FirestoreEvent.isFavorite() = false // TODO add actual favourites filtering
+
+    private fun Observable<List<FirestoreSchedulePage>>.filterByTracks(selectedTracks: Observable<Set<Track>>) =
+        Observable.combineLatest(
+            this,
+            selectedTracks,
+            BiFunction { pages: List<FirestoreSchedulePage>, tracks: Set<Track> ->
+                pages.filterPagesEvents { it.hasTrackOrNoTrack(tracks) }
+            }
         )
-    }
 
-    private fun <T, U, V> combineInATriple(): Function3<List<T>, U, Set<V>, Triple<List<T>, U, Set<V>>> = Function3(::Triple)
+    private fun List<FirestoreSchedulePage>.filterPagesEvents(predicate: (FirestoreEvent) -> Boolean) =
+        map { it.filterPageEvents(predicate) }
 
-    private fun List<Event>.filterOnlyFavorites(onlyFavorites: Boolean) = when {
-        onlyFavorites -> filter { it.favorited }
-        else -> this
+    private fun FirestoreSchedulePage.filterPageEvents(predicate: (FirestoreEvent) -> Boolean): FirestoreSchedulePage {
+        val filteredEvents = events.filter(predicate)
+        return FirestoreSchedulePage().apply {
+            events = filteredEvents
+            day = this@filterPageEvents.day
+        }
     }
 
-    private fun List<Event>.filterByTracks(tracks: Set<Track>) = when {
-        tracks.isEmpty() -> this
-        else -> filter { shouldIncludeEvent(it, tracks) }
-    }
+    private fun FirestoreEvent.hasTrackOrNoTrack(allowedTracks: Set<Track>) =
+        track?.let { allowedTracks.any { it.id == this.track?.id ?: it.id } } ?: true
```

Comment:
instead of using `this.track?.id` (you know it's there), add a name to the lambda parameter:
```kotlin
track?.let { eventTrack -> allowedTracks.any { it.id == eventTrack.id } } ?: true }
```